File: java/Graphs/BidirectionalGraph.java
package Graphs;

import definitions.Graphs;

import java.util.*;

public class BidirectionalGraph<T> implements Graphs<T> {

    private final Map<T, Set<T>> adjMap;

    public BidirectionalGraph() {
        adjMap = new HashMap<>();
    }

    @Override
    public void addEdge(T source, T destination) {
        if (!adjMap.containsKey(source)) adjMap.put(source, new HashSet<>());
        if (!adjMap.containsKey(destination)) adjMap.put(destination, new HashSet<>());
        adjMap.get(source).add(destination);
        adjMap.get(destination).add(source);
    }

    @Override
    public void removeEdge(T source, T destination) {
        if (!adjMap.containsKey(source) || !adjMap.containsKey(destination)) return;
        adjMap.get(source).remove(destination);
        adjMap.get(destination).remove(source);
    }

    @Override
    public List<List<T>> breadthFirstSearch() {
        List<List<T>> ans = new LinkedList<>();
        Set<T> vis = new HashSet<>();
        adjMap.forEach((k, v) -> {
            if (!vis.contains(k))
                ans.add(breadthFirstSearch(k, vis));
        });
        return ans;
    }

    @Override
    public List<T> breadthFirstSearch(T source) {
        return breadthFirstSearch(source, new HashSet<>());
    }

    private List<T> breadthFirstSearch(T source, Set<T> vis) {
        List<T> ans = new LinkedList<>();
        if (!adjMap.containsKey(source)) return ans;
        Queue<T> q = new LinkedList<>();
        q.add(source);
        while (!q.isEmpty()) {
            var node = q.remove();
            vis.add(node);
            adjMap.get(node).forEach(adjNode -> {
                if (!vis.contains(adjNode)) q.add(adjNode);
            });
            ans.add(node);
        }
        return ans;
    }

    @Override
    public List<List<T>> depthFirstSearch() {
        Set<T> vis = new HashSet<>();
        List<List<T>> ans = new LinkedList<>();
        adjMap.forEach((k, v) -> {
            if (!vis.contains(k)) ans.add(depthFirstSearch(k, vis));
        });
        return ans;
    }

    @Override
    public List<T> depthFirstSearch(T source) {
        return depthFirstSearch(source, new HashSet<>());
    }

    private List<T> depthFirstSearch(T source, Set<T> vis) {
        List<T> ans = new LinkedList<>();
        if (!adjMap.containsKey(source)) return ans;
        vis.add(source);
        ans.add(source);
        adjMap.get(source).forEach(adjNode -> {
            if (!vis.contains(adjNode)) {
                ans.addAll(depthFirstSearch(adjNode, vis));
            }
        });
        return ans;
    }

    @Override
    public String toString() {
        return this.adjMap.toString();
    }

}

File: typescript/src/graphs/bi-directional-graph.ts
import Queue from "../queues/queue.ts";

export default class BiDirectionalGraph<T> {
  private adjMap: Map<T, Set<T>>;

  public constructor() {
    this.adjMap = new Map();
  }

  public addEdge(src: T, dest?: T) {
    if (!this.adjMap.has(src)) this.adjMap.set(src, new Set());
    if (!dest) return;
    if (!this.adjMap.has(dest)) this.adjMap.set(dest, new Set());
    this.adjMap.get(src)!.add(dest);
    this.adjMap.get(dest)!.add(src);
  }

  public removeEdge(src: T, dest?: T) {
    if (!dest) {
      this.adjMap.get(src)?.forEach((adjNode) => {
        this.adjMap.get(adjNode)?.delete(src);
      });
      this.adjMap.delete(src);
      return;
    }
    this.adjMap.get(src)?.delete(dest);
  }

  public bfsOfNode(src: T) {
    return this.bfsHelper(src, new Set());
  }

  private bfsHelper(src: T, vis: Set<T>) {
    const ans: T[] = [];
    const q = new Queue<T>();
    q.add(src);
    while (q.size > 0) {
      const node = q.remove()!;
      vis.add(node);
      ans.push(node);
      this.adjMap.get(node)!.forEach((adjNode) => {
        if (!vis.has(adjNode)) q.add(adjNode);
      });
    }
    return ans;
  }

  public bfs(): T[][] {
    const ans: T[][] = [];
    const vis = new Set<T>();
    this.adjMap.forEach((_, k) => {
      if (!vis.has(k)) ans.push(this.bfsHelper(k, vis));
    });
    return ans;
  }

  public dfs() {
    const ans: T[][] = [];
    const vis = new Set<T>();
    this.adjMap.forEach((_, k) => {
      if (!vis.has(k)) ans.push(this.dfsHelper(k, vis));
    });
    return ans;
  }

  public dfsOfNode(src: T) {
    if (!this.adjMap.has(src)) return [];
    return this.dfsHelper(src, new Set());
  }

  private dfsHelper(src: T, vis: Set<T>): T[] {
    const ans: T[] = [];
    ans.push(src);
    vis.add(src);
    this.adjMap.get(src)?.forEach((adjNode) => {
      if (!vis.has(adjNode)) {
        ans.push(...this.dfsHelper(adjNode, vis));
      }
    });
    return ans;
  }

  public toString() {
    return this.adjMap.entries().toArray().map(([key, val]) => {
      return [key, "->[" + [...val].join(",") + "]"];
    }).join(", ");
  }

  public get vertices() {
    return this.adjMap.size;
  }
}
